xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java

```diff
@@ -745,6 +1002,138 @@ void onRejected(String rejectedVersion, long rejectedTime, String rejectedDetail
     private void notifyWatcher(ResourceWatcher<T> watcher, T update) {
       watcher.onChanged(update);
     }
+
+    void setCpc(ControlPlaneClient controlPlaneClient) {
+      checkNotNull(controlPlaneClient, "controlPlaneClient");
+      this.controlPlaneClient = controlPlaneClient;
+    }
+  }
+
+  private class ResponseHandler implements XdsResponseHandler {
+    final ServerInfo serverInfo;
+
+    ResponseHandler(ServerInfo serverInfo) {
+      this.serverInfo = serverInfo;
+    }
+
+    @Override
+    public void handleResourceResponse(
+        XdsResourceType<?> xdsResourceType, ServerInfo serverInfo, String versionInfo,
+        List<Any> resources, String nonce, ProcessingTracker processingTracker) {
+      checkNotNull(xdsResourceType, "xdsResourceType");
+      syncContext.throwIfNotInThisSynchronizationContext();
+      Set<String> toParseResourceNames =
+          xdsResourceType.shouldRetrieveResourceKeysForArgs()
+          ? getResourceKeys(xdsResourceType)
+          : null;
+      XdsResourceType.Args args = new XdsResourceType.Args(serverInfo, versionInfo, nonce,
+          bootstrapInfo, securityConfig, toParseResourceNames);
+      handleResourceUpdate(args, resources, xdsResourceType, processingTracker);
+    }
+
+    @Override
+    public void handleStreamClosed(Status status, boolean inRetry,
+                                   ControlPlaneClient cpcThatClosed) {
+      syncContext.throwIfNotInThisSynchronizationContext();
+      ControlPlaneClient cpcForThisStream = serverCpClientMap.get(serverInfo);
+      if (!inRetry) {
+        cleanUpResourceTimers(cpcForThisStream);
+      }
+
+      Status error = status.isOk() ? Status.UNAVAILABLE.withDescription(CLOSED_BY_SERVER) : status;
+      boolean checkForFallback = !status.isOk() && BootstrapperImpl.isEnabledXdsFallback()
+          && !inRetry;
+
+      for (Map<String, ResourceSubscriber<? extends ResourceUpdate>> subscriberMap :
+          resourceSubscribers.values()) {
+        for (ResourceSubscriber<? extends ResourceUpdate> subscriber : subscriberMap.values()) {
+          if (subscriber.hasResult() || subscriber.controlPlaneClient != cpcThatClosed
+              || (subscriber.serverInfos != null && !subscriber.serverInfos.contains(serverInfo))) {
```

Comment:
If `subscriber.controlPlaneClient != cpcThatClosed` is false, then the third operand is always false. So the third operand can be removed, otherwise it's hard to reason about.

Reply:
done